lib/timeline-window.js

```diff
@@ -0,0 +1,392 @@
+
+    var delta = this._eventCount - this._windowLimit;
+
+    while (delta > 0) {
+        var count = startOfTimeline ? tl.advance(delta) : tl.retreat(delta);
+        if (count <= 0) {
+            // sadness. This shouldn't be possible.
+            throw new Error(
+                "Unable to unpaginate any further, but still have " +
+                    this._eventCount + " events");
+        }
+
+        delta -= count;
+        this._eventCount -= count;
+        debuglog("TimelineWindow.unpaginate: dropped " + count +
+                 " (now " + this._eventCount + ")");
+    }
+};
+
+
+/**
+ * Get a list of the events currently in the window
+ *
+ * @return {MatrixEvent[]} the events in the window
+ */
+TimelineWindow.prototype.getEvents = function() {
+    if (!this._start) {
+        // not yet loaded
+        return [];
+    }
+
+    var result = [];
+
+    for (var timeline = this._start.timeline; timeline !== null;
+        timeline = timeline.getNeighbouringTimeline(EventTimeline.FORWARDS)) {
+
+        var events = timeline.getEvents();
+        var startIndex = 0, endIndex = events.length;
+        if (timeline === this._start.timeline) {
+            startIndex = this._start.index + timeline.getBaseIndex();
+        }
+        if (timeline === this._end.timeline) {
+            endIndex = this._end.index + timeline.getBaseIndex();
+        }
+
+        for (var i = startIndex; i < endIndex; i++) {
+            result.push(events[i]);
+        }
+    }
+    return result;
+};
+
+
+/**
+ * a thing which contains a timeline reference, and an index into it.
+ *
+ * @constructor
+ * @param {EventTimeline} timeline
+ * @param {number} index
+ * @private
+ */
+function TimelineIndex(timeline, index) {
+    this.timeline = timeline;
+
+    // the indexes are relative to BaseIndex, so could well be negative.
+    this.index = index;
+}
+
+/**
+ * @return {number} the minimum possible value for the index in the current
+ *    timeline
+ */
+TimelineIndex.prototype.minIndex = function() {
+    return -this.timeline.getBaseIndex();
+};
+
+/**
+ * @return {number} the maximum possible value for the index in the current
+ *    timeline
+ */
+TimelineIndex.prototype.maxIndex = function() {
+    return this.timeline.getEvents().length - this.timeline.getBaseIndex();
+};
+
+/**
+ * Try move the index forward, or into the neighbouring timeline
+ *
+ * @param {number} delta  number of events to advance by
+ * @return {number} number of events successfully advanced by
+ */
+TimelineIndex.prototype.advance = function(delta) {
+    // first try moving the cap
+    var x;
+    if (delta === 0) {
+        return 0;
+    } else if (delta < 0) {
+        x = Math.max(delta, this.minIndex() - this.index);
+        if (x < 0) {
+            this.index += x;
+            return x;
+        }
+    } else {
+        x = Math.min(delta, this.maxIndex() - this.index);
+        if (x > 0) {
+            this.index += x;
+            return x;
```

Comment:
Please explain what you're doing, and I would prefer a most useful var name than `x`. I'm particularly interested in the case where `delta > 0` and yet `x < 0` which would mean it would fall-through. Intentional? Possible to happen?